### PR TITLE
Benchmark for smf

### PR DESCRIPTION
**Pull request by albertlast**
Something I like to wish since start of my dev,
a way the benchmark smf.

for different motivation:
- changes on database stuff (dev stuff)
- changes on hardware (change webhoster)
- changes on software releases (change of php or database version)

the second and last point i believe it's a common issue and not special like the first one.

Two modes exists:
User creation
Post creation

Both takes a minute and try to create as mutch as possible.

### CODE REVIEW

**Comment by jdarwood007** [comment]
Should note that registerMember is a really really slow function.  I believe during my testing it came down to the password creation as hash_password was designed to be slow.

**Comment by albertlast** [comment]
Well this in my eyes no problem.

I did some test in my env.

mysql
-user 81
-topic 72

postgres
-user 112
-topic 7789

Dunno why the numbers with pg so mutch higher than mysql (i believe my mysql is not so good).
But it's shows that the user creation is mutch more cpu/php heavy task is and
the topic run more database releated is.

And this in my eyes a good mix.

**Comment by albertlast** [comment]
Added a third methode,
this try to access a topic x times in one minutes.

This benchmark shows realy good how many topic a user can access,
because this stuff depend on many things like size of the board and activate features.
It's show realy good the status of the board,
but this make it's hard for comparing accross with different board.

Because it's show how a mod, feature or size effect the board,
make this test veary valuable for board admins.

Also I try to add in the description what the test focus on.

File: Sources/ManageMaintenance.php [hunk 1]
```diff
@@ -2285,4 +2295,257 @@ function get_hook_info_from_raw($rawData)
 	return $hookData;
 }
 
+/**
+ * Benchmakr for User creation tries to create as many as possible in 1 minute
```
Comment (colinschoen) [inline Sources/ManageMaintenance.php]: `Benchmakr` -> `benchmark`

File: Sources/ManageMaintenance.php [hunk 2]
```diff
@@ -2285,4 +2295,257 @@ function get_hook_info_from_raw($rawData)
...
+	removeTopics($postid);
+	
+	$context['benchmark_result']['amount'] = $count;
+	$context['benchmark_result']['test_name'] = $txt['benchmark_post'];
+	
+	deleteMembers($userID);
+}
+
+/**
+ * Benchmakr for Post reads tries to access as many as possible in 1 minute
```
Comment (colinschoen) [inline Sources/ManageMaintenance.php]: `Benchmakr` -> `Benchmark`

File: Themes/default/languages/ManageMaintenance.english.php [hunk 3]
```diff
@@ -230,4 +230,13 @@
 $txt['enable_maintenance1'] = 'Due to the size of your forum, it is recommended to place your forum in "maintenance mode" before you start the backup.';
 $txt['enable_maintenance2'] = 'To proceed, due to the size of your forum, please place your forum in "maintenance mode".';
 
+$txt['benchmark_usercreate'] = 'Creating User';
+$txt['benchmark_post'] = 'Creating Posts';
+$txt['benchmark_usercreate_info'] = 'Create as many users as possible in one minute <br/> This test is cpu heavy';
+$txt['benchmark_post_info'] = 'Create as many posts as possible in one minute <br/> This test is db insert imperformance heavy';
```
Comment (colinschoen) [inline Themes/default/languages/ManageMaintenance.english.php]: `imperformance` -> `performance`

File: Themes/default/languages/ManageMaintenance.english.php [hunk 4]
```diff
@@ -230,4 +230,13 @@
 $txt['enable_maintenance1'] = 'Due to the size of your forum, it is recommended to place your forum in "maintenance mode" before you start the backup.';
 $txt['enable_maintenance2'] = 'To proceed, due to the size of your forum, please place your forum in "maintenance mode".';
 
+$txt['benchmark_usercreate'] = 'Creating User';
+$txt['benchmark_post'] = 'Creating Posts';
+$txt['benchmark_usercreate_info'] = 'Create as many users as possible in one minute <br/> This test is cpu heavy';
+$txt['benchmark_post_info'] = 'Create as many posts as possible in one minute <br/> This test is db insert imperformance heavy'; 
+$txt['benchmark_topic'] = 'Benchmark methodes';
```
Comment (colinschoen) [inline Themes/default/languages/ManageMaintenance.english.php]: `Benchmark methodes` -> `Benchmark methods`

File: Themes/default/languages/ManageMaintenance.english.php [hunk 5]
```diff
@@ -230,4 +230,13 @@
...
+$txt['benchmark_postread'] = 'Reading Posts';
+$txt['benchmark_postread_info'] = 'Access a topic how often it can in one minute <br/> This test dipence heavy on the size of your board and active features';
```
Comment (colinschoen) [inline Themes/default/languages/ManageMaintenance.english.php]: `Repeatedly access a topic for 1 minute. <br /> This test depends heavily on the size of your board and the active features.`

**Review by colinschoen — CHANGES_REQUESTED** [review]
This benchmarking has some side effects of affecting statistics and increasing unique IDs substantially if run often.

**Comment by albertlast** [comment]
The statistics are not touched as fare i see in my test env.
I see no problem with raising the unique id's in db tables,
max int is 2147483647 in my test was max 8k for topic creating stuff...

**Comment by jdarwood007** [comment]
Would this only have use for development/testing?

I'm not seeing this as something that is meant for a production site.  Perhaps it would be best split off into a separate utility that we could expand and maintain as part of benchmark and/or unit testing

**Comment by albertlast** [comment]
i already sayed my point of in the first commit,
in short i see need on admin side too.

**Comment by XinYenFon** [comment]
I disagree this is a bloated addition to core.

**Comment by Sesquipedalian** [comment]
This kind of benchmarking is definitely useful, @albertlast. But I agree with the others that it would be better to develop it as a separate mod or utility. That would give you more flexibility to adjust the various types of benchmarks and add new ones over time. By the look of things, you should be able to do everything via integration hooks.

**Comment by albertlast** [comment]
I take from the last comments basicly: "The dev do the thing" 
Can someone show me the person
how test smf with all variation of php version are supported,
all databases types and version,
different sizes of boards,
different smf configuration (cache setting/mode etc.),
with the different combination of mods and
hardware/webspace configuration which is smf running?

When you can not mention this guy than is my eye not provided by the dev "guy" and
the smf admin had to do this stuff.

**Comment by Sesquipedalian** [comment]
What we are saying is that if you make it a mod for this, then anyone who feels the need to do some benchmarking can just install the mod and do as much as they like. It should be pretty easy to make a mod out of this, and in fact you should be able to do it entirely using integration hooks.

**Comment by albertlast** [comment]
But i will create no mod,
this type of information are essential,
when this type is not essential we should drop function like calendar too.

**Comment by XinYenFon** [comment]
What? How you correlate calendar with benchmark ?

**Comment by colinschoen** [comment]
The issue is that 99% of forum owners won't have too much use for the utility. It is however extremely useful for the 1%, but that means it is likely better as a mod to install for those that will use it.

**Comment by XinYenFon** [comment]
On the other hand, you can tweak https://github.com/SimpleMachines/tools/blob/master/Populate.php to do benchmark as well as the populate or create whole new file like populate (benchmark.php) to make it happen.

**Comment by albertlast** [comment]
Populate is the wrong tool to benchmark your smf,
please don't use it for this case...

**Comment by Sesquipedalian** [comment]
Using the `integrate_admin_areas` and `integrate_manage_maintenance` hooks, you should be able to make a mod for this that does everything you want, @albertlast. The consensus seems to be that this would be the best approach to follow.

**Comment by albertlast** [comment]
When you mean this benchmark would be good for dev,
why you not keeping this and activate only when $db_show_debug=true; is on.

From my side will be no mod.

**Comment by colinschoen** [comment]
The issue isn't that it is hidden or visible, but that it inflates the code base which we have to maintain for a feature that will likely be used by ~1% of the users. What is your hesitate to convert this into a mod?

**Comment by albertlast** [comment]
Why you not create a mod the code is there?
Easy anwser the issue the support.

**Comment by colinschoen** [comment]
^ Sorry, I don't understand. Can you rephrase.

**Comment by albertlast** [comment]
provide a mod create many support task and i'm not willig to do so,
when you mean that is important than feel free to take the code an create mod.

**Comment by colinschoen** [comment]
Sure thing I will make the mod.

**Comment by jdarwood007** [comment]
We can spin up a new repo to contain the mod call it smf benchmark.  Contain it in there as a mod.  As a plus, allow us to easily call it from a benchmark.php as cron/cli.  So then us devs could pull down the mod and latest 2.1 dev, then have the cron script go through, running the benchmark and sending the results out (either in post or email).  As a plus, that benchmark.php could then be a git hook as well :)

**Comment by albertlast** [comment]
@colinschoen what is the status of the mod?

**Comment by colinschoen** [comment]
Low on the priority list. Planned but not started.

**Comment by albertlast** [comment]
I was interesting how mod creation works in smf,
so i take this pr and created a mod: https://custom.simplemachines.org/mods/index.php?mod=4155
It's nice on the one side and on the other side it's pain.
